# EDA

In [40]:
import pandas as pd
import numpy as np

In [41]:
# Example importing the CSV here

df = pd.read_csv('../data/raw/AB_NYC_2019.csv')


In [64]:
df.sample(10, random_state=1234)

id                                               name    host_id  \
36514  29034190                             The  NYC Getaway Space    9808458   
48599  36334813            Sunny Room in the heart of Williamsburg   17379828   
34181  27110361                      Artist's home in Williamsburg    8731758   
21356  17013882   Big bedroom in Astoria, 20 min from Central Park    7039858   
4887    3493833     Huge Artist Loft with BBQ deck -  WILLIAMSBURG   17592620   
42954  33362377  Bright, Williamsburg bedroom with private balcony    4094038   
43907  33937844    Spacious, Light-Filled, Private Bklyn 2-Bedroom  255990804   
16372  13162150                 Sunfilled, clean, MONTHly discount    4326776   
39504  30768969      Sunny & Big Room w/private toilet in Bushwick  192181166   
22661  18340498    Private bedroom and bathroom near Prospect Park   33723491   

      host_name neighbourhood_group  neighbourhood  latitude  longitude  \
36514      Zain               Bronx    Kingsbridge  40.87072  -73.89974   
48599  Caterina            Brooklyn   Williamsburg  40.70906  -73.95544   
34181    Jeanne            Brooklyn   Williamsburg  40.71326  -73.96440   
21356    Costas              Queens        Astoria  40.76882  -73.92793   
4887       Paul            Brooklyn   Williamsburg  40.71788  -73.96021   
42954      Alex            Brooklyn   Williamsburg  40.71157  -73.94055   
43907      Joel            Brooklyn  Crown Heights  40.66549  -73.94616   
16372    Sabina              Queens       Woodside  40.74457  -73.90474   
39504     Elisa            Brooklyn   Williamsburg  40.70891  -73.94268   
22661  Suzzanne            Brooklyn       Flatbush  40.65274  -73.95848   

             room_type  price  minimum_nights  number_of_reviews last_review  \
36514     Private room     65               2                  0         NaT   
48599     Private room     75               3                  0         NaT   
34181  Entire home/apt    175               5                  5  2019-05-23   
21356     Private room     65               3                 23  2018-12-31   
4887   Entire home/apt    318               5                 19  2019-06-01   
42954     Private room     75               2                  4  2019-06-04   
43907  Entire home/apt    220               2                  4  2019-06-05   
16372     Private room     80              14                  0         NaT   
39504     Private room     72               3                  0         NaT   
22661     Private room     39               2                  7  2017-09-29   

       reviews_per_month  calculated_host_listings_count  availability_365  
36514                NaN                               4               364  
48599                NaN                               1                 7  
34181               0.46                               1               198  
21356               0.81                               1                23  
4887                0.32                               1               268  
42954               2.11                               4                 0  
43907               2.03                               1               251  
16372                NaN                               1                 0  
39504                NaN                               2                 0  
22661               0.27                               1                 0

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

La base de datos tiene 48.895 observaciones y 16 variables, de las cuales cuatro tienen valores nulos. Por ahora como no sabemos qué variables vamos a utilizar, no se realizará ningún tratamiento de los null.

También se observa que existen tres variables del tipo float64, 7 del tipo int 64 y 6 de tipo object. Como algunas variables se importaron en un tipo de dato incorrecto, antes de comenzar con el procesamiento de la base de datos se transformarán al tipo de datos más adecuado.

In [65]:
df=df.astype({'name':'str','host_name':'str','neighbourhood_group':'category','neighbourhood':'category','room_type':'category'})
df['last_review'] = pd.to_datetime(df['last_review'], format="%Y/%m/%d")
df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                     category
neighbourhood                           category
latitude                                 float64
longitude                                float64
room_type                               category
price                                      int64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
dtype: object

A continuación se estudian algunas estadísticas resumen sobre las variables para tener una primer idea de la composición de las variables.

In [61]:
round(df.describe())

id      host_id  latitude  longitude    price  minimum_nights  \
count     48895.0      48895.0   48895.0    48895.0  48895.0         48895.0   
mean   19017143.0   67620011.0      41.0      -74.0    153.0             7.0   
std    10983108.0   78610967.0       0.0        0.0    240.0            21.0   
min        2539.0       2438.0      40.0      -74.0      0.0             1.0   
25%     9471945.0    7822033.0      41.0      -74.0     69.0             1.0   
50%    19677284.0   30793816.0      41.0      -74.0    106.0             3.0   
75%    29152178.0  107434423.0      41.0      -74.0    175.0             5.0   
max    36487245.0  274321313.0      41.0      -74.0  10000.0          1250.0   

       number_of_reviews  reviews_per_month  calculated_host_listings_count  \
count            48895.0            38843.0                         48895.0   
mean                23.0                1.0                             7.0   
std                 45.0                2.0                            33.0   
min                  0.0                0.0                             1.0   
25%                  1.0                0.0                             1.0   
50%                  5.0                1.0                             1.0   
75%                 24.0                2.0                             2.0   
max                629.0               58.0                           327.0   

       availability_365  
count           48895.0  
mean              113.0  
std               132.0  
min                 0.0  
25%                 0.0  
50%                45.0  
75%               227.0  
max               365.0

calculated_host_listings_count = cantidad de inmuebles del propietario
max noches 1250??
precio 0_??

In [46]:
df.describe(include=['object','category','datetime64'])

/tmp/ipykernel_1430/1335291268.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df.describe(include=['object','category','datetime64'])


name host_name neighbourhood_group neighbourhood  \
count            48895     48895               48895         48895   
unique           47906     11453                   5           221   
top     Hillside Hotel   Michael           Manhattan  Williamsburg   
freq                18       417               21661          3920   
first              NaN       NaN                 NaN           NaN   
last               NaN       NaN                 NaN           NaN   

              room_type          last_review  
count             48895                38843  
unique                3                 1764  
top     Entire home/apt  2019-06-23 00:00:00  
freq              25409                 1413  
first               NaN  2011-03-28 00:00:00  
last                NaN  2019-07-08 00:00:00

In [53]:
print('Cantidad de hosts:', df['host_id'].nunique())

Cantidad de hosts: 37457


In [56]:
print('Tipos de alojamientos')
df['room_type'].value_counts(normalize=True)

Tipos de alojamientos


Entire home/apt    0.519665
Private room       0.456611
Shared room        0.023724
Name: room_type, dtype: float64

De la descripción anterior se obtiene que existen 37.457 host que ofrecen 47.905 alojamientos, por lo que podemos intuir que algún host administra varios alojamientos. Se observa también que los alojamientos se ubican en 221 barrios que se agrupan en 5 regiones. 

Con respecto a las características de los alojamientos se puede decir que existen tres tipos de 'espacios' ofrecidos, siendo el alojamiento entero el más ofrecido.

### What can we learn about different hosts and areas?

SyntaxError: invalid syntax (1728185032.py, line 1)

In [59]:
df['host_id'].value_counts(sort=True)[:10] # Top 10 host

219517861    327
107434423    232
30283594     121
137358866    103
16098958      96
12243051      96
61391963      91
22541573      87
200380610     65
7503643       52
Name: host_id, dtype: int64

In [80]:
#df.groupby('host_id')[['number_of_reviews','id']].agg({'id':pd.Series.nunique,'number_of_reviews':['sum','mean']}).sort_values(by=('number_of_reviews','mean'),ascending=False)[0:10] # Top 10 host
df.groupby('host_id')[['number_of_reviews','id']].agg({'id':pd.Series.nunique,'number_of_reviews':['sum','mean']}).sort_values(by=('id','nunique'),ascending=False)[0:10] # Top 10 host

id number_of_reviews          
          nunique               sum      mean
host_id                                      
219517861     327              1281  3.917431
107434423     232                29  0.125000
30283594      121                65  0.537190
137358866     103                87  0.844660
16098958       96               138  1.437500
12243051       96                43  0.447917
61391963       91               417  4.582418
22541573       87                55  0.632184
200380610      65                 1  0.015385
1475015        52               162  3.115385

In [114]:
df_hn=round(df.groupby(['host_id','neighbourhood_group'])[['number_of_reviews','id','price']].agg({'id':pd.Series.nunique,'number_of_reviews':['sum','mean'],'price':'mean'}).sort_values(by=('id','nunique'),ascending=False))[0:10] # Top 10 host
df_hn

id number_of_reviews       price
                              nunique               sum mean   mean
host_id   neighbourhood_group                                      
219517861 Manhattan             327.0              1281  4.0  253.0
107434423 Manhattan             230.0                29  0.0  303.0
30283594  Manhattan             121.0                65  1.0  278.0
12243051  Manhattan              96.0                43  0.0  213.0
16098958  Manhattan              96.0               138  1.0  209.0
61391963  Manhattan              91.0               417  5.0  146.0
22541573  Manhattan              86.0                54  1.0  216.0
137358866 Queens                 79.0                54  1.0   44.0
200380610 Manhattan              65.0                 1  0.0  290.0
1475015   Manhattan              52.0               162  3.0  103.0

In [97]:
list(df_hn.reset_index()['host_id'])

[219517861,
 107434423,
 30283594,
 12243051,
 16098958,
 61391963,
 22541573,
 137358866,
 200380610,
 1475015]

In [101]:
df_hn_filt=df[df['host_id'].isin(list(df_hn.reset_index()['host_id']))]
df_hn_filt

id                                               name    host_id  \
1386     613556                2 BED TriBeCa, Beautiful-Renovated!    1475015   
1454     652371                ELEGANT MIDTOWN EAST STUDIO E.52 ST    1475015   
1455     652466  32nd St & Lexington Ave / Doorman Beautiful St...    1475015   
1456     652515                 COLUMBUS CIRCLE~FULLY FURNISHED!!!    1475015   
1457     652648                GRAMERCY PARK~FURNISHED E.20's ST P    1475015   
...         ...                                                ...        ...   
48723  36404684  Smart Nolita 1BR w/ W/D near famous cafes by B...  107434423   
48724  36404784  Dapper Hell's Kitchen 2BR w/  Gym + W/D + Door...  107434423   
48725  36404815  Trendy Hell's Kitchen 1BR w/ Gym + W/D + Doorm...  107434423   
48726  36404936  Beautiful Williamsburg 2BR w/ Skyline views + ...  107434423   
48727  36404972  Hip East Village 1BR w/ Gym, W/D, Doorman, nea...  107434423   

        host_name neighbourhood_group   neighbourhood  latitude  longitude  \
1386         Mike           Manhattan         Tribeca  40.71655  -74.01171   
1454         Mike           Manhattan         Midtown  40.75743  -73.96939   
1455         Mike           Manhattan        Kips Bay  40.74190  -73.98160   
1456         Mike           Manhattan  Hell's Kitchen  40.76758  -73.98722   
1457         Mike           Manhattan        Kips Bay  40.74189  -73.97833   
...           ...                 ...             ...       ...        ...   
48723  Blueground           Manhattan          Nolita  40.72283  -73.99472   
48724  Blueground           Manhattan  Hell's Kitchen  40.76082  -73.99709   
48725  Blueground           Manhattan  Hell's Kitchen  40.76083  -73.99727   
48726  Blueground            Brooklyn    Williamsburg  40.71493  -73.96365   
48727  Blueground           Manhattan    East Village  40.73206  -73.98776   

             room_type  price  minimum_nights  number_of_reviews last_review  \
1386   Entire home/apt    130              30                  2  2015-10-31   
1454   Entire home/apt     90              30                  3  2016-12-31   
1455   Entire home/apt    100              30                  2  2018-06-30   
1456   Entire home/apt     85              30                  2  2016-08-15   
1457   Entire home/apt     87              30                  3  2018-08-15   
...                ...    ...             ...                ...         ...   
48723  Entire home/apt    316              30                  0         NaT   
48724  Entire home/apt    385              30                  0         NaT   
48725  Entire home/apt    267              30                  0         NaT   
48726  Entire home/apt    278              30                  0         NaT   
48727  Entire home/apt    365              30                  0         NaT   

       reviews_per_month  calculated_host_listings_count  availability_365  
1386                0.03                              52               116  
1454                0.05                              52               358  
1455                0.04                              52               342  
1456                0.04                              52               223  
1457                0.05                              52               363  
...                  ...                             ...               ...  
48723                NaN                             232               325  
48724                NaN                             232               338  
48725                NaN                             232               228  
48726                NaN                             232               188  
48727                NaN                             232               295  

[1270 rows x 16 columns]

In [116]:
pd.set_option('display.max_rows',110)
round(df_hn_filt.groupby(['host_id','neighbourhood_group','neighbourhood'])[['number_of_reviews','id','price']].agg({'id':pd.Series.nunique,'number_of_reviews':['sum','mean'],'price':'mean'}).sort_values(by=['host_id',('id','nunique')],ascending=False).dropna()) # Top 10 host

id number_of_reviews  \
                                                 nunique               sum   
host_id   neighbourhood_group neighbourhood                                  
219517861 Manhattan           Financial District   218.0              1129   
                              Murray Hill           50.0               141   
                              Theater District      27.0                 6   
                              Hell's Kitchen        15.0                 3   
                              Chelsea                7.0                 0   
                              Upper East Side        6.0                 2   
                              Midtown                4.0                 0   
200380610 Manhattan           Midtown               17.0                 0   
                              Upper East Side       10.0                 0   
                              Hell's Kitchen         9.0                 0   
                              Upper West Side        6.0                 0   
                              Chelsea                5.0                 1   
                              Murray Hill            5.0                 0   
                              Theater District       4.0                 0   
                              East Village           2.0                 0   
                              Gramercy               2.0                 0   
                              Kips Bay               2.0                 0   
                              West Village           2.0                 0   
                              Battery Park City      1.0                 0   
137358866 Queens              Woodside              35.0                22   
                              Sunnyside             17.0                14   
          Manhattan           Harlem                12.0                18   
          Queens              Elmhurst              10.0                11   
                              Astoria                9.0                 2   
          Brooklyn            Bushwick               6.0                 6   
          Manhattan           East Harlem            6.0                 9   
          Queens              Jackson Heights        4.0                 3   
                              Maspeth                3.0                 2   
                              Long Island City       1.0                 0   
107434423 Manhattan           Chelsea               37.0                 7   
                              Tribeca               30.0                 3   
                              Theater District      27.0                 2   
                              Financial District    25.0                 4   
                              Upper West Side       17.0                 2   
                              East Village          14.0                 1   
                              Kips Bay              14.0                 1   
                              West Village          14.0                 1   
                              Midtown               13.0                 1   
                              Hell's Kitchen        10.0                 1   
                              Upper East Side       10.0                 3   
                              Murray Hill            9.0                 2   
                              Battery Park City      2.0                 1   
                              Flatiron District      2.0                 0   
                              Nolita                 2.0                 0   
          Brooklyn            DUMBO                  1.0                 0   
                              Williamsburg           1.0                 0   
          Manhattan           Gramercy               1.0                 0   
                              Lower East Side        1.0                 0   
                              NoHo                   1.0                 0  

In [111]:
df[['neighbourhood_group','neighbourhood']].value_counts(sort=True)[:10]

neighbourhood_group  neighbourhood     
Brooklyn             Williamsburg          3920
                     Bedford-Stuyvesant    3714
Manhattan            Harlem                2658
Brooklyn             Bushwick              2465
Manhattan            Upper West Side       1971
                     Hell's Kitchen        1958
                     East Village          1853
                     Upper East Side       1798
Brooklyn             Crown Heights         1564
Manhattan            Midtown               1545
dtype: int64

In [123]:
df_nei=df.groupby(['neighbourhood_group','neighbourhood']).agg({'price':'mean'}).sort_values(by=['neighbourhood_group','price'],ascending=False).dropna()

In [126]:
df_nei.reset_index(inplace=True)


In [128]:
list(df_nei.neighbourhood_group.unique())

['Staten Island', 'Queens', 'Manhattan', 'Brooklyn', 'Bronx']

In [130]:
for barrio in list(df_nei.neighbourhood_group.unique()) :
    display(df_nei[df_nei['neighbourhood_group'] == barrio].head(5))

Staten Island


neighbourhood_group   neighbourhood  price
0       Staten Island  Fort Wadsworth  800.0
1       Staten Island         Woodrow  700.0
2       Staten Island    Prince's Bay  409.5
3       Staten Island   Randall Manor  336.0
4       Staten Island     Willowbrook  249.0

Queens


neighbourhood_group    neighbourhood       price
43              Queens         Neponsit  274.666667
44              Queens     Breezy Point  213.333333
45              Queens  Jamaica Estates  182.947368
46              Queens          Arverne  171.779221
47              Queens     Belle Harbor  171.500000

Manhattan


neighbourhood_group      neighbourhood       price
94           Manhattan            Tribeca  490.638418
95           Manhattan  Battery Park City  367.557143
96           Manhattan  Flatiron District  341.925000
97           Manhattan               NoHo  295.717949
98           Manhattan               SoHo  287.103352

Brooklyn


neighbourhood_group     neighbourhood       price
126            Brooklyn          Sea Gate  487.857143
127            Brooklyn       Cobble Hill  211.929293
128            Brooklyn  Brooklyn Heights  209.064935
129            Brooklyn             DUMBO  196.305556
130            Brooklyn      Vinegar Hill  187.176471

Bronx


neighbourhood_group   neighbourhood       price
173               Bronx       Riverdale  442.090909
174               Bronx     City Island  173.000000
175               Bronx  Spuyten Duyvil  154.750000
176               Bronx     Eastchester  141.692308
177               Bronx       Unionport  137.142857

In [131]:
for barrio in list(df_nei.neighbourhood_group.unique()) :
    display(df_nei[df_nei['neighbourhood_group'] == barrio].tail(5))

neighbourhood_group   neighbourhood      price
38       Staten Island         Concord  58.192308
39       Staten Island      Grant City  57.666667
40       Staten Island  New Dorp Beach  57.400000
41       Staten Island        New Dorp  57.000000
42       Staten Island     Bull's Head  47.333333

neighbourhood_group neighbourhood      price
89              Queens     Ridgewood  77.184397
90              Queens      Rosedale  76.694915
91              Queens   Little Neck  75.200000
92              Queens     Woodhaven  67.170455
93              Queens        Corona  59.171875

neighbourhood_group        neighbourhood       price
121           Manhattan  Morningside Heights  114.783237
122           Manhattan     Roosevelt Island  113.259740
123           Manhattan   Washington Heights   89.610679
124           Manhattan          Marble Hill   89.166667
125           Manhattan               Inwood   88.896825

neighbourhood_group neighbourhood      price
168            Brooklyn       Midwood  80.339450
169            Brooklyn     Gravesend  79.014706
170            Brooklyn   Brownsville  76.459016
171            Brooklyn   Bensonhurst  75.786667
172            Brooklyn  Borough Park  63.066176

neighbourhood_group neighbourhood      price
216               Bronx    Mount Eden  58.500000
217               Bronx     Bronxdale  57.105263
218               Bronx     Soundview  53.466667
219               Bronx       Tremont  51.545455
220               Bronx   Hunts Point  50.500000